### ISSUE TITLE
Race condition between loginAPIEndpoint request and appPageURL redirect on onIdTokenChange event

### ISSUE THREAD

**Issue by hershmire** [gladly-team/next-firebase-auth#189]
**Describe the bug**
I'm noticing a race condition happening where the `loginAPIEndpoint` API fetch call doesn't always complete before the application redirects the user to the `appPageURL` location resulting in the fetch call getting cancelled. This only appears to be an issue if the `appPageURL` is pointing to a page that has a global redirect set in `next.config.js`. 

What seems to be happening is the `onIdTokenChange` event will fire and [set the firebase `user` and `initialized` React state before the awaited `setAuthCookie()` function is called](https://github.com/gladly-team/next-firebase-auth/blob/main/src/useFirebaseUser.js#L75-L77). Since the state changes before the async fetch request is executed, the [`withAuthUser()`](https://github.com/gladly-team/next-firebase-auth/blob/main/src/withAuthUser.js) HOC can re-render too early and start redirecting the application without waiting on a successful login to occur. 

If the login API request is cancelled, cookies are never set so we end up in a half broken state where the user is authenticated on the client-side but the server never knows resulting in any SSR authenticated pages continuing to redirect back to the login page.

**Version**
What version of `next-firebase-auth` are you using?
0.13.1

**To Reproduce**
I've created this PR to demo how to reproduce the issue: https://github.com/gladly-team/next-firebase-auth/pull/188

NOTE: This issue is only reproducible when running the application in production mode.

**Expected behavior**
When a user logs in, no redirection should occur until the auth cookies are set.

**Comment by kmjennison** [comment]
Thanks for the clear description and demo.

As a fix, it would work to wait for the auth cookie call before setting state in `useFirebaseUser`:
```js
await setAuthCookie(firebaseUser) // moved up
setUser(firebaseUser)
setInitialized(true)
```

However, that would require the app to wait for a server call _every time_ before the auth is initialized, which would be a bad performance hit.

A better approach would be to return [another property from `useFirebaseUser`](https://github.com/gladly-team/next-firebase-auth/blob/main/src/useFirebaseUser.js#L89), like `setAuthCookieComplete`, then wait for that to be true before [redirecting to the app](https://github.com/gladly-team/next-firebase-auth/blob/main/src/withAuthUser.js#L77). This would wait for the server call only on pages that should redirect when the user is authenticated. Note that the client won't know for sure if the cookies were in fact set, given they're HTTP-only.

Do you think that would fix this?

**Comment by hershmire** [comment]
> However, that would require the app to wait for a server call every time before the auth is initialized, which would be a bad performance hit.

Ah, because that call is happening each time this library is initialized?

> A better approach would be to return another property from useFirebaseUser, like setAuthCookieComplete, then wait for that to be true before redirecting to the app. This would wait for the server call only on pages that should redirect when the user is authenticated. Note that the client won't know for sure if the cookies were in fact set, given they're HTTP-only.

If I'm understanding correctly, internally you would wait to redirect the app based on the value from `setAuthCookieComplete` so I think that would work as long as that variable is set once the login request finishes. The cookie should be set if the login API request is successful. The demo I provided should help prove that out.

**Comment by kmjennison** [comment]
> Ah, because that call is happening each time this library is initialized?

Yes, + whenever the Firebase ID token changes.

> If I'm understanding correctly, internally you would wait to redirect the app based on the value from `setAuthCookieComplete` so I think that would work as long as that variable is set once the login request finishes.

Yes, I think we're on the same page! PR welcome for this if you have the time.

**Comment by davidtsai** [comment]
I'm having a similar issue to this where my app re-renders with user authenticated on the client before the server side cookie is set. This causes my app to fire off API calls where the user is unauthenticated even though they are already authenticated on the client. Should we be following a different pattern in order to prevent this?

**Comment by kmjennison** [comment]
Released in [v0.13.3-alpha.0](https://github.com/gladly-team/next-firebase-auth/releases/tag/v0.13.3-alpha.0). Please let me know if you run into any problems!
